source/agora/consensus/validation/Block.d

```diff
@@ -136,7 +140,31 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             return fail_reason;
     }
 
-    return null;
+    return validateBlockTimestamp(prev_timestamp, block.header.timestamp, curr_timestamp, blockTimestampToleranceSecs);
+}
+
+static public string validateBlockTimestamp(ulong prev_block_ts, ulong new_block_ts, ulong curr_timestamp,
+        uint blockTimestampToleranceSecs) @safe nothrow
+{
+    string res;
+    try
+    {
+        import std.conv : to;
+        import std.datetime.systime;
+        if (new_block_ts <= prev_block_ts)
+            res = "the timestamp in consensus data is not greater than the timestamp in the last block\n" ~
+                    "new block timestamp: " ~ to!string(SysTime(unixTimeToStdTime(new_block_ts))) ~ "\n" ~
+                    "previous block's timestamp: " ~ to!string(SysTime(unixTimeToStdTime(prev_block_ts)));
+        else if (new_block_ts > curr_timestamp + blockTimestampToleranceSecs)
+            res =  "(the timestamp in consensus data) is greater than (the timestamp in the last block + tolerance)\n" ~
+                    "new block timestamp: " ~ to!string(SysTime(unixTimeToStdTime(new_block_ts))) ~ "\n" ~
+                    "current timestamp: " ~ to!string(SysTime(unixTimeToStdTime(curr_timestamp))) ~ "\n" ~
+                    "tolerance in seconds: " ~to!string(blockTimestampToleranceSecs);
+    }
+    catch (Exception e) {}
```

Comment:
If the `try` block throws without setting `res`, this would return valid here. Should `res` be set to something in the `catch` block?

Reply:
yeah, forgot to remove that, just did it for testing, thanks